packages/go_router/lib/src/route.dart

```diff
@@ -446,22 +473,201 @@ class ShellRoute extends RouteBase {
+///               builder: (BuildContext context, GoRouterState state) =>
+///               const DetailsScreen(label: 'A'),
+///             ),
+///           ],
+///         ),
+///
+///         /// The root of tab 'B'
+///         GoRoute(
+///           path: '/b',
+///           builder: (BuildContext context, GoRouterState state) =>
+///           const RootScreen(label: 'B', detailsPath: '/b/details/1'),
+///           routes: <RouteBase>[
+///             /// Will cover screen B but not the bottom navigation bar
+///             GoRoute(
+///               path: 'details',
+///               builder: (BuildContext context, GoRouterState state) =>
+///               const DetailsScreen(label: 'B'),
+///             ),
+///           ],
+///         ),
+///       ],
+///     ),
+///   ],
+/// );
+/// ```
+///
+/// For scenarios where it's suitable to use an [IndexedStack] to manage the
+/// navigators, consider using the [PartitionedShellRoute.stackedNavigationShell]
+/// constructor instead to reduce boilerplate. For example:
+///
+/// ```
+/// final GoRouter _router = GoRouter(
+///   initialLocation: '/a',
+///   routes: <RouteBase>[
+///     PartitionedShellRoute.stackedNavigation(
+///       stackItems: [
+///         StackedNavigationItem(
+///           rootRoutePath: '/a', navigatorKey: _tabANavigatorKey),
+///         StackedNavigationItem(
+///           rootRoutePath: '/b', navigatorKey: _tabBNavigatorKey),
+///       ],
+///       scaffoldBuilder: (BuildContext context, int currentIndex,
+///           List<StackedNavigationItemState> itemsState,
+///           Widget scaffoldBody) {
+///         return ScaffoldWithBottomNavigationBar(
+///             currentIndex: currentIndex,
+///             itemsState: itemsState,
+///             body: scaffoldBody,
+///         );
+///       },
+///       routes: <RouteBase>[
+///         /// The root of tab 'A'
+///         GoRoute(
+///           path: '/a',
+///           builder: (BuildContext context, GoRouterState state) =>
+///             const RootScreen(label: 'A', detailsPath: '/a/details'),
+///         ),
+///         /// The root of tab 'B'
+///         GoRoute(
+///           path: '/b',
+///           builder: (BuildContext context, GoRouterState state) =>
+///             const RootScreen(label: 'B', detailsPath: '/b/details/1'),
+///         ),
+///       ],
+///     ),
+///   ],
+/// );
+/// ```
+///
+/// See [Stateful Nested Navigation](https://github.com/flutter/packages/blob/main/packages/go_router/example/lib/stateful_nested_navigation.dart)
+/// for a complete runnable example.
+class PartitionedShellRoute extends ShellRouteBase {
+  /// Constructs a [PartitionedShellRoute] from a list of routes, each
+  /// representing the root of a navigation hierarchy. A separate
+  /// [Navigator] will be created for each of the root routes, using the
+  /// navigator key at the corresponding index in [navigatorKeys].
+  PartitionedShellRoute({
```

Comment:
I think people who use this class will only use stackedNavigationShell anyway. I suggest we make it the only constructor to use this class.